stdlib/src/builtin/string.mojo

```diff
@@ -2109,3 +2205,83 @@ fn _calc_format_buffer_size[type: DType]() -> Int:
         return 64 + 1
     else:
         return 128 + 1  # Add 1 for the terminator
+
+
+# ===----------------------------------------------------------------------===#
+# Format method structures
+# ===----------------------------------------------------------------------===#
+
+
+@value
+struct _FormatCurlyEntry:
+    """
+    Internally used by the `format()` method.
+
+    Purpose of the variant types:
+
+    - `String` for **kwargs
+
+        ```mojo
+        "{message}".format(message="hello")
+        ```
+
+        value field contains `message`
+
+    - `Int` for manual indexing of *args
+
+        ```mojo
+        "{0}".format("hello")
+        ```
+
+        value field contains `0`
+
+    - `NoneType` for automatic indexing of *args
+
+        ```mojo
+        "{}".format("hello")
+        ```
+
+        value field contains `None`
+
+    `*_curly` are the indexes of the enclosing curlies
+
+    This struct only supports theses:
+
+    `{number|name|nothing}`
+
+    """
+
+    var first_curly: Int
+    var last_curly: Int
+    var value: Variant[String, Int, NoneType]
+
+    @staticmethod
+    fn create_entries(arg: String) -> List[Self]:
+        """Used internally by the `format()` method.
+
+        Args:
+            arg: The "format" part provided by the user.
+
+        Returns:
+            A `List` of structured format entries.
+        """
+        var Entries = List[Self]()
+        var start = Optional[Int](None)
+        for i in range(len(arg)):
+            if arg[i] == "{":
+                start = i
+                continue
+            if arg[i] == "}":
+                if start:
+                    var start_value = start.value()[]
+                    var tmp = Self(start_value, i, None)
+                    if i - start_value != 1:
+                        var tmp2 = arg[start_value + 1 : i]
+                        try:
+                            var tmp3 = int(tmp2)
+                            tmp.value = tmp3
```

Comment:
What's raising here and what kind of exception are we trying to catch?

Reply:
If the `int` raises it means it is a keyword for **kwargs (`{name}`),
if it does not raises, it is a number that can be used for manual indexing (`{0} {1}`)
If the slice is len(0), it is `NoneType`, for automatic indexing (`{} {}`)

I think we need thoses 3, even if not using kwargs, in order to raise errors and ensure python behaviour;

Reply:
Okay, thanks for explaining these! Can you please put these in a comment here? Also, would it be possible to handle these cases in some more explicit way? We don't have typed errors yet, so exception-as-control-flow is not as expressive as in Python.

Reply:
Yes ! Did some improvements, we might need to do another review of that part